lib/_stream_writable.js

```diff
@@ -489,83 +473,69 @@ function afterWrite(stream, state, count, cb) {
 
 // If there's something in the buffer waiting, then invoke callbacks.
 function errorBuffer(state, err) {
-  if (state.writing || !state.bufferedRequest) {
+  if (state.writing) {
     return;
   }
 
-  for (let entry = state.bufferedRequest; entry; entry = entry.next) {
-    const len = state.objectMode ? 1 : entry.chunk.length;
+  for (let n = state.bufferedIndex; n < state.buffered.length; ++n) {
+    const { chunk, callback } = state.buffered[n];
+    const len = state.objectMode ? 1 : chunk.length;
     state.length -= len;
-    entry.callback(err);
+    callback(err);
   }
-  state.bufferedRequest = null;
-  state.lastBufferedRequest = null;
-  state.bufferedRequestCount = 0;
+
+  resetBuffer(state);
 }
 
 // If there's something in the buffer waiting, then process it
 function clearBuffer(stream, state) {
+  if (state.corked || state.bufferProcessing) {
+    return;
+  }
+
+  const { buffered, bufferedIndex, objectMode } = state;
+  const bufferedLength = buffered.length - bufferedIndex;
+
+  if (!bufferedLength) {
+    return;
+  }
+
+  let i = bufferedIndex;
+
   state.bufferProcessing = true;
-  let entry = state.bufferedRequest;
-
-  if (stream._writev && entry && entry.next) {
-    // Fast case, write everything using _writev()
-    const l = state.bufferedRequestCount;
-    const buffer = new Array(l);
-    const holder = state.corkedRequestsFree;
-    holder.entry = entry;
-
-    let count = 0;
-    let allBuffers = true;
-    while (entry) {
-      buffer[count] = entry;
-      if (entry.encoding !== 'buffer')
-        allBuffers = false;
-      entry = entry.next;
-      count += 1;
-    }
-    buffer.allBuffers = allBuffers;
+  if (bufferedLength > 1 && stream._writev) {
+    state.pendingcb -= bufferedLength - 1;
+
+    const callback = state.allNoop ? nop : (err) => {
+      for (let n = i; n < buffered.length; ++n) {
+        buffered[n].callback(err);
+      }
+    };
+    // Make a copy of `buffered` if it's going to be used by `callback` above,
+    // since `doWrite` will mutate the array.
+    const chunks = state.allNoop && i === 0 ? buffered : buffered.slice(i);
```

Comment:
@mafintosh: I don't think FixedQueue will help us. Part of the optimization here is that we can  send the buffer *as is* to `doWrite`. Without having to create a new copy.

Reply:
To optimise writev?

Reply:
yes

Reply:
I'm actually more interested in making this easier to maintain. So I'm open for `FixedQueue`, the performance won't be the same thoughl

Reply:
FixedQueue will likely make _write faster and your approach _writev. We actually know at creation time which one the user is using so we could impl a hybrid at some point.